src/OrchardCore.Modules/OrchardCore.Contents/Sitemaps/ContentTypesSitemapUpdateHandler.cs

```diff
@@ -1,36 +1,49 @@
+using System;
 using System.Threading.Tasks;
 using OrchardCore.ContentManagement.Handlers;
+using OrchardCore.Locking.Distributed;
 using OrchardCore.Sitemaps.Handlers;
 
 namespace OrchardCore.Contents.Sitemaps
 {
     public class ContentTypesSitemapUpdateHandler : ContentHandlerBase
     {
+        private readonly IDistributedLock _distributedLock;
         private readonly ISitemapUpdateHandler _sitemapUpdateHandler;
 
-        public ContentTypesSitemapUpdateHandler(ISitemapUpdateHandler sitemapUpdateHandler)
+        public ContentTypesSitemapUpdateHandler(IDistributedLock distributedLock, ISitemapUpdateHandler sitemapUpdateHandler)
         {
+            _distributedLock = distributedLock;
             _sitemapUpdateHandler = sitemapUpdateHandler;
         }
 
-        public override Task PublishedAsync(PublishContentContext context)
-        {
-            var updateContext = new SitemapUpdateContext
-            {
-                UpdateObject = context.ContentItem,
-            };
+        public override Task PublishedAsync(PublishContentContext context) => UpdateSitemapDeferredAsync(context);
 
-            return _sitemapUpdateHandler.UpdateSitemapAsync(updateContext);
-        }
+        public override Task UnpublishedAsync(PublishContentContext context) => UpdateSitemapDeferredAsync(context);
 
-        public override Task UnpublishedAsync(PublishContentContext context)
+        // Doing the update in a synchronized way makes sure that two simultaneous content item updates don't cause a
+        // ConcurrencyException due to the same sitemap document being updated.
+        private async Task UpdateSitemapDeferredAsync(ContentContextBase context)
         {
-            var updateContext = new SitemapUpdateContext
+            var timeout = TimeSpan.FromMilliseconds(20_000);
+            (var locker, var locked) = await _distributedLock.TryAcquireLockAsync("SITEMAPS_UPDATE_LOCK", timeout, timeout);
```

Comment:
Please move all the locking logic into the `DefaultSitemapUpdateHandler` instead of this content handler.

Reply:
Yeah, that's better. I originally wanted a deferred task, but that proved to be pointless, but that should've been in the handler. BTW moving this locking any deeper wouldn't work due to how `SitemapManager` returns a mutable object.